cmd/goal/application.go

```diff
@@ -995,3 +1050,152 @@ var infoAppCmd = &cobra.Command{
 		}
 	},
 }
+
+var methodAppCmd = &cobra.Command{
+	Use:     "method",
+	Short:   "Invoke a method",
+	Long:    `Invoke a method in an App (stateful contract) with an application call transaction`,
+	Args:    validateNoPosArgsFn,
+	PreRunE: validateNoPosArgsFn,
+	Run: func(cmd *cobra.Command, args []string) {
+		var applicationArgs [][]byte
+
+		// insert the method selector hash
+		hash := sha512.Sum512_256([]byte(method))
+		applicationArgs = append(applicationArgs, hash[0:4])
+
+		// parse down the ABI type from method signature
+		argTupleTypeStr, retTypeStr, err := abi.ParseMethodSignature(method)
+		if err != nil {
+			reportErrorf("cannot parse method signature: %v", err)
+		}
+		err = abi.ParseArgJSONtoByteSlice(argTupleTypeStr, methodArgs, &applicationArgs)
+		if err != nil {
+			reportErrorf("cannot parse arguments to ABI encoding: %v", err)
+		}
+
+		// Parse transaction parameters
+		appArgsParsed, appAccounts, foreignApps, foreignAssets := getAppInputs()
+		if len(appArgsParsed) > 0 {
+			reportErrorf("in goal app method: --arg and --app-arg are mutually exclusive, do not use --app-arg")
+		}
+		if len(appAccounts) > 0 {
+			reportErrorf("in goal app method: --app-account is not supported")
+		}
+		if len(foreignApps) > 0 {
+			reportErrorf("in goal app method: --foreign-app is not supported")
+		}
+		if len(foreignAssets) > 0 {
+			reportErrorf("in goal app method: --foreign-asset is not supported")
```

Comment:
I see, so it might be okay to unconstrain `foreignApps`, `foreignAssets` and `appAccounts` for now?
I suppose `appArgs` should be mutual exclusive with `--arg` in `goal app method`, so I will still disable that.

Reply:
Assuming there was support for the special types `account`/`application`/`asset`, does it make sense to also support the `--foreign-*` flags? I think not, for the same reason we don't allow you to also specify additional raw byte arguments with `--app-arg`.

And I believe we should support the special types `account`/`application`/`asset` soon in a follow up PR. So while it's regrettable that foreign properties won't be settable immediately, I'd rather not allow the `--foreign-*` flags temporarily only to remove them (or worse, forget to remove them) later.

Reply:
Oh I see you've already removed them @ahangsu. That's fine, I don't care enough about this to ask you to add them back. We can just address it in the follow up PR

Reply:
These feel different to me.  The ABI describes exactly how ApplicationArgs are to be used. So it does not make sense to allow people to set them using --app-arg, and maybe mess up what the ABI expects.  But the ABI does not specify how your foreign arrays will be packed. It merely says that IF you use a special type as an argument, it'll go (somewhere!) in the appropriate foreign-array, and the index will be encoded.

It's perfectly ok to let the caller put extra resources into the foreign arrays.  And I think it will be useful as people adopt the ABI to greater or lesser extents. I think I can even imagine that users will *need* this, even after ABI is complete and totally adopted, but I need to think more.  At any rate, the main point is that the ABI doesn't say how the foreign arrays will be packed, but does say how the application args will be packed.

Reply:
I suppose this also means we might be clever about encoding the foreign arrays.  If the same account is used as two different arguments, it can be put in once, and both arguments should refer to it. (I don't care a lot if we do this.)

Reply:
Good point about the ABI not specifying exactly where the foreign arguments will show up. In light of that, allowing additional foreign params sounds fine to me.

Reply:
Just asking, so do we want to allow --app-arg here anyways? Or it is just fine for now?

Reply:
I vote that it's fine for now